mozilla-services/socorro
# bug 1753521: refactor indexing and start migrating some raw crash fields

This refactors indexing to make it more flexible around moving data around. Also, it starts the move to my ultimate vision where analytics and data flow rely solely on processed (normalized, validated, redacted, etc) data rather than the original values from the crash annotations in the crash report.

Towards that aim, this also starts migrating indexing from pulling from the raw crash and indexing into the `raw_crash` namespace to pull from the processed crash and indexing into the `processed_crash` namespace. We're going to do a slow migration because that's all we have support for at the moment. In 5 months, we'll remove the raw crash sides of these things.

There's a handful of followup bugs from this work to cover issues I bumped into and decided to carve out as separate bugs.

After this lands, we still have to migrate the boolean-ish fields that are being pulled from the raw crash. We have to figure out how to process those and unify how we index them--right now we do a variety of different things and it's a right mess.

To test this, I did a few things:

1. Generate a mapping using the `main` branch (`socorro-cmd es list` and then `socorro-cmd es print_mapping`) and compare with the mapping generated from this branch. The `raw_crash` namespace should be identical between the two mappings. The `processed_crash` namespace has a bunch of new fields in it.
2. Process crash reports with the main branch and capture their Elasticsearch document sources (`curl "http://elasticsearch:9200/${INDEX}/crash_reports/${CRASHID}" > ${CRASHID}.document.json`) and then switch to this branch, process the same crash reports, capture their Elasticsearch document sources, and compare the two.

   `raw_crash.UptimeTS` switched from a string value to a float. It wasn't being converted correctly in the old indexing code which didn't handle "double".

   `raw_crash.ContentSandboxCapabilities` switched from a string value to a short. It also wasn't being converted correctly in the old indexing code which didn't handle "short".

   There should be a bunch of additional data in the `processed_crash` namespace--that's the data we're migrating over.
3. Supersearch should continue to work--both querying and faceting.
4. Signature generation should be based mostly on the processed crash data (I think I have one more field to fix--that's covered in another bug). While signature generation looks at some different fields for some data, signatures shouldn't change with these changes.

The one thing I didn't check is `TelemetryBotoS3CrashStorage` data. I'm pretty sure none of these changes affect how that's generated. I wrote up another bug to rework that so it's not using `super_search_fields.py` to build its crash data.
